app/services/atmosphere/billing_service.rb

```diff
@@ -131,12 +140,25 @@ def self.can_afford_vm?(appliance, vm)
       if appliance.fund.blank?
         raise Atmosphere::BillingException.new(message: "can_afford_vm? invoked on an appliance (with id #{appliance.id}) which has no fund assigned. Unable to proceed.")
       end
-      billable_time = self.calculate_billable_time(appliance, Time.now.utc, true)
-      amt_due = (billable_time*(vm.virtual_machine_flavor.hourly_cost)/(vm.appliances.count+1)).round
+
+      # Check whether the target VM is already assigned to the appliance via deployments
+      # If so, respect the prepayment period which *may* be present for this deployment
+      deployments = appliance.deployments.select {|dep| dep.virtual_machine == vm}
+      if deployments.count > 0
+        billable_time = self.calculate_billable_time(deployments.first, Time.now.utc, true)
+      else
+        # If this VM is not assigned to appliance then just assume def prepayment interval
+        billable_time = @appliance_prepayment_interval/3600 # Time in hours
+      end
+      hourly_cost = vm.virtual_machine_flavor.get_hourly_cost_for(appliance.appliance_type.os_family)
+      if hourly_cost.blank? # Will happen when there is a mismatch between AT ostype and flavor ostype
+        return false
+      end
+      amt_due = ((billable_time*hourly_cost)/(vm.appliances.count+1)).round
```

Comment:
The +1 represents the (still hypothetical) new appliance for which the charge is returned. Basically, the VM is shared by n appliances and we're asking how much it would cost to assign another appliance to it.